polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+                }
+                assert!(fields.fields.is_empty());
+                let var = match lhs.value() {
+                    Value::Variable(var) | Value::RestVariable(var) => var.clone(),
+                    Value::Expression(op) if op.operator == Operator::Dot => {
+                        dot_var(var_info, op.args[0].clone(), &op.args[1])
+                    }
+                    _ => todo!(),
+                };
+                var_info.types.push((var, tag.clone().0))
+            } else {
+                unimplemented!("Non pattern specializers are not yet supported for data filtering.")
+            }
+        }
+        Operator::Unify | Operator::Eq | Operator::Assign => {
+            assert_eq!(exp.args.len(), 2);
+
+            let mut lhs = exp.args[0].clone();
+            if let Value::Expression(op) = lhs.value() {
+                lhs = process_exp(var_info, op).unwrap();
+            };
+
+            let mut rhs = exp.args[1].clone();
+            if let Value::Expression(op) = rhs.value() {
+                rhs = process_exp(var_info, op).unwrap();
+            };
+
+            match (lhs.value(), rhs.value()) {
+                // Unifying two variables
+                (Value::Variable(l), Value::Variable(r)) => {
+                    var_info.cycles.push((l.clone(), r.clone()));
+                }
+                // Unifying a variable with a value
+                (Value::Variable(var), val) | (val, Value::Variable(var)) => var_info
+                    .eq_values
+                    .push((var.clone(), Term::new_temporary(val.clone()))),
+                // Unifying something else.
+                // 1 = 1 is irrelevant for data filtering, other stuff seems like an error.
+                // @NOTE(steve): Going with the same not yet supported message but if this is
+                // coming through it's probably a bug in the simplifier.
+                (_a, _b) => {
+                    unimplemented!(
+                        "Unification of values is not yet supported for data filtering."
+                    );
+                }
+            };
+        }
+        Operator::In => {
+            assert_eq!(exp.args.len(), 2);
+
+            let mut lhs = exp.args[0].clone();
+            if let Value::Expression(op) = lhs.value() {
+                lhs = process_exp(var_info, op).unwrap();
+            };
+
+            let mut rhs = exp.args[1].clone();
+            if let Value::Expression(op) = rhs.value() {
+                rhs = process_exp(var_info, op).unwrap();
+            };
+
+            match (lhs.value(), rhs.value()) {
+                // l in r
+                (Value::Variable(l), Value::Variable(r)) => {
+                    var_info.in_relationships.push((l.clone(), r.clone()));
+                }
+                // var in [1, 2, 3]
+                (Value::Variable(_var), _val) => {
```

Comment:
I don't think this every happens because VM would produce choice points & multiple bindings.

Reply:
Right